sdk/iot/provisioning/src/az_iot_provisioning_client.c

```diff
@@ -47,161 +54,420 @@ AZ_NODISCARD az_result az_iot_provisioning_client_init(
+  az_span mqtt_topic_filter_span
+      = az_span_init((uint8_t*)mqtt_topic_filter, (int32_t)mqtt_topic_filter_size);
+  int32_t required_length = az_span_size(str_dps_registrations_res) + 1;
 
-  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_filter, required_length);
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(mqtt_topic_filter_span, required_length);
 
-  az_span remainder = az_span_copy(mqtt_topic_filter, str_dps);
-  remainder = az_span_copy(remainder, str_registrations);
-  remainder = az_span_copy(remainder, str_res);
-  az_span_copy_u8(remainder, '#');
+  az_span remainder = az_span_copy(mqtt_topic_filter_span, str_dps_registrations_res);
+  remainder = az_span_copy_u8(remainder, '#');
+  remainder = az_span_copy_u8(remainder, '\0');
 
-  *out_mqtt_topic_filter = az_span_slice(mqtt_topic_filter, 0, required_length);
+  if (out_mqtt_topic_filter_length)
+  {
+    *out_mqtt_topic_filter_length = (size_t)required_length;
+  }
 
   return AZ_OK;
 }
 
-// topic: $dps/registrations/res/202/?$rid=1&retry-after=3
-// payload:
-// {"operationId":"4.d0a671905ea5b2c8.42d78160-4c78-479e-8be7-61d5e55dac0d","status":"assigning"}
-AZ_NODISCARD az_result az_iot_provisioning_client_received_topic_payload_parse(
+AZ_INLINE az_iot_provisioning_client_registration_state registration_state_default()
+{
+  return (az_iot_provisioning_client_registration_state){ .assigned_hub_hostname = AZ_SPAN_NULL,
+                                                          .device_id = AZ_SPAN_NULL,
+                                                          .status = AZ_IOT_STATUS_OK,
+                                                          .extended_error_code = 0,
+                                                          .error_message = AZ_SPAN_NULL,
+                                                          .error_tracking_id = AZ_SPAN_NULL,
+                                                          .error_timestamp = AZ_SPAN_NULL };
+}
+
+/*
+Documented at
+https://docs.microsoft.com/en-us/rest/api/iot-dps/runtimeregistration/registerdevice#deviceregistrationresult
+  "registrationState":{
+    "x509":{},
+    "registrationId":"paho-sample-device1",
+    "createdDateTimeUtc":"2020-04-10T03:11:13.0276997Z",
+    "assignedHub":"contoso.azure-devices.net",
+    "deviceId":"paho-sample-device1",
+    "status":"assigned",
+    "substatus":"initialAssignment",
+    "lastUpdatedDateTimeUtc":"2020-04-10T03:11:13.2096201Z",
+    "etag":"IjYxMDA4ZDQ2LTAwMDAtMDEwMC0wMDAwLTVlOGZlM2QxMDAwMCI="}}
+*/
+AZ_INLINE az_result az_iot_provisioning_client_payload_error_code_parse(
+    az_json_token_member* tm,
+    az_iot_provisioning_client_registration_state* out_state)
+{
+  if (az_span_is_content_equal(AZ_SPAN_FROM_STR("errorCode"), tm->name))
+  {
+    double value;
+    AZ_RETURN_IF_FAILED(az_json_token_get_number(&tm->token, &value));
+    out_state->extended_error_code = (uint32_t)value;
+    out_state->status = (uint32_t)value / 1000;
```

Comment:
What is in `az_iot_status` and why is this division by 1000 ok? What does it represent?

Reply:
`az_iot_status` and `az_iot_provisioning_client_registration_state.extended_error_code` are public API and have documentation.

```C
// (from enum) @brief Azure IoT service status codes.

  az_iot_status status; /**< The register operation status. */
  uint32_t extended_error_code; /**< The extended, 6 digit error code. */
```

Reply:
Right, but none of this mentioned anything about division by 1000 and that there is some inherent relationship between status and extended_error_code. Are you relying on existing domain/service knowledge of the reader (which I don't have much of)? Then that's fine, but as someone maintaining the code, I will come back to it months from now and ask, why is this 1000 and what happens if I change it to 100?

Reply:
Makes sense. This should actually be a common function used by other services like Twin. 
@danewalton @ewertons - we'll need to unify extended-error to `az_iot_status` transformation. My proposal is to have an inline internal within az_iot_core.

@ahsonkhan I'll follow-up in another PR.

Reply:
I've added `_az_iot_status_from_extended_status` within az_iot_provisioning_client.c 
I couldn't find any need to promote this to an internal within az_iot_core as Twin is not parsing any errors from payload (yet). 

If we need to have the 6 digit to 3 digit transformation done anywhere else, this should be moved to az_iot_core.h as an internal (or public) API. 

@danewalton @ewertons  @jspaith FYI /cc @hihigupt @ericwol-msft